src/Pipeline/Atomic.php

```diff
@@ -62,56 +64,82 @@ protected function getConnection()
     protected function executePipeline(ConnectionInterface $connection, \SplQueue $commands)
     {
         $profile = $this->getClient()->getProfile();
-        $connection->executeCommand($profile->createCommand('multi'));
+        $multi = $profile->createCommand('MULTI');
+        $exec = $profile->createCommand('EXEC');
 
+        // open transaction: MULTI
+        $connection->writeRequest($multi);
+
+        // transmit all commands in transaction
         foreach ($commands as $command) {
             $connection->writeRequest($command);
         }
 
+        // execute transaction: EXEC
+        $connection->writeRequest($exec);
+
+        // response to MULTI
+        $response = $connection->readResponse($multi);
+        if ($response instanceof ErrorResponseInterface) {
+            $this->exception($connection, $response); // close connection and throw ServerException
+        }
+
+        // responses to commands (all should be QUEUED result)
+        $ex = null;
         foreach ($commands as $command) {
             $response = $connection->readResponse($command);
 
-            if ($response instanceof ErrorResponseInterface) {
-                $connection->executeCommand($profile->createCommand('discard'));
-                throw new ServerException($response->getMessage());
+            // throw first error as a ServerException
+            if (($response instanceof ErrorResponseInterface) && !$ex) {
+                $ex = new ServerException($response->getMessage());
             }
         }
 
-        $executed = $connection->executeCommand($profile->createCommand('exec'));
+        // response to EXEC
+        $responses = $connection->readResponse($exec);
 
-        if (!isset($executed)) {
-            // TODO: should be throwing a more appropriate exception.
-            throw new ClientException(
-                'The underlying transaction has been aborted by the server.'
-            );
+        // if one of the commands returned an error, only throw after all responses have been received
+        if ($ex) {
+            throw $ex;
         }
 
-        if (count($executed) !== count($commands)) {
+        // if EXEC returned null, the transaction was aborted
+        if (!isset($responses)) {
+            // fake a MultiExec, which is not used in an atomic pipeline
+            $multi_exec = new MultiExec($this->getClient());
+            throw new AbortedMultiExecException($multi_exec, 'The underlying transaction has been aborted by the server.');
```

Comment:
I do understand the logic behind this as a specific exception would be ideal in this context, but I don't like it anyway as you are throwing an exception which expects a certain type as an argument and you work around this by instantiating a fake object (with an internal state that potentially differs from the current state of the pipelined transaction) that has no meaning when intercepted by outside code in a try / catch block.

Would it be possible to go with a generic `Predis\ClientException` for now in v1.1?

In v2.0 we may consider having `Predis\Transaction\AbortedMultiExecException` accept a `Predis\ClientContextInterface` as an argument for `__construct()` (pipelines and plain transactions implement this interface after all) and we could even add a specific interface like `Predis\Transaction\TransactionInterface` if we can find some specific common ground that could work with both pipelined and non-pipelined transactions, or eventually just make it a marker interface that implements `Predis\ClientContextInterface`.

Reply:
Trust me, I don't like it either, but this is pretty important to me.  We have systems relying on `WATCH/MULTI/EXEC` to perform certain operations, and we keep stats on which operations failed due to errors (which is a problem to be fixed) versus those which failed due to concurrency issues (which is part of the design).  This change permitted us to (finally) distinguish between the two cases without performing string comparisons of exception messages.  (It also meant we could migrate plain `transaction()` code to atomic pipelines without further code changes.)

If making this change is that troublesome, I'll revert.  I filed a ticket for this at #640.

Reply:
The description of your use case makes perfect sense so we'll fix #640 for sure in v2.0, thanks for creating a separate issue so that we can keep track of it.

Back to this on the stable branch, just thinking out loud... if we really want to change I'd probably prefer having a new exception extending `Predis\Transaction\AbortedMultiExecException` and implemented with something along these lines (I'm skipping phpdocs just for the sake of brevity):

```php
namespace Predis\Pipeline;

use Predis\Transaction\AbortedMultiExecException;

class AbortedAtomicPipelineException extends AbortedMultiExecException
{
    public function __construct($message, $code = null)
    {
        parent::__construct(null, $message, $code);
    }

    public function getTransaction()
    {
        return null;
    }
}
```

This would make transactions and atomic pipelines somewhat interchangeable when aborted transaction errors are returned by Redis as you'd just need to catch `Predis\Transaction\AbortedMultiExecException`. Furthermore, I'd vastly prefer `getTransaction()` to return `NULL` instead of an unrelated instance of `Predis\Transaction\MultiExecTransaction`. Then again, my concern is that there would be no way to retrieve the original instance of `Predis\Client` in a `catch` block. Hmmmm....

PS: generally speaking, this can be indeed considered a bug but would such a change be considered backwards-compatible for a patch release?

Reply:
I would be ecstatic to add a subclass for the exception, but as you pointed out, there are other concerns there with the base class contract.  Personally, I'm not overly concerned about retrieving the `Client` in the `catch` block, but I can understand if other callers have structured their code around that.

> this can be indeed considered a bug but would such a change be considered backwards-compatible for a patch release?

That's ... a good point.  If there is existing code out there expecting a `ClientException` for the abort, then there would be some compatibility issues.  But this would be a good change for v2.0, since those assumptions can be revised.  Or, for v1.x, there could be a field stored in the client's `OptionsInterface` we could add signalling "throw AbortedMultiExecException in atomic pipelines" is allowed (and make it mandatory for v2.0.)

Reply:
Reverted.  Now throws `ClientException` as before.